Add response writer and publisher

SbiCA opened the pull request:
Adds support for a prometheus response writer and gateway publisher.

@unaizorrilla  I'm not sure regarding the projects could you please let me know if they are fine.

Cheers

unaizorrilla commented:
Hi @SbiCA 

I'm sorry for delay on check this PR, I try to review soon

unaizorrilla commented on src/HealthChecks.Publisher.Prometheus/DependencyInjection/PrometheusGatewayHealthCheckBuilderExtensions.cs:
```diff
@@ -0,0 +1,38 @@
+            string endpoint, string job, string instance = null)
+        {
+            builder.Services
+                .AddSingleton<IHealthCheckPublisher>(sp => new PrometheusGatewayPublisher(endpoint, job,instance));
```
Fix spaces on parameters

unaizorrilla commented on src/HealthChecks.Publisher.Prometheus/DependencyInjection/PrometheusGatewayHealthCheckBuilderExtensions.cs:
```diff
@@ -0,0 +1,38 @@
+        public static IHealthChecksBuilder AddPrometheusGatewayPublisher(this IHealthChecksBuilder builder, HttpClient client, string endpoint, string job, string instance = null)
+        {
+            builder.Services
+                .AddSingleton<IHealthCheckPublisher>(sp => new PrometheusGatewayPublisher(client,endpoint, job,instance));
```
fix space on parameters

unaizorrilla commented on src/HealthChecks.Publisher.Prometheus/PrometheusGatewayPublisher.cs:
```diff
@@ -0,0 +1,66 @@
+        public PrometheusGatewayPublisher(string endpoint, string job, string instance = null)
+        {
+            var sb = new StringBuilder($"{endpoint.TrimEnd('/')}/job/{job}");
+            if (!string.IsNullOrEmpty(instance)) sb.AppendFormat("/instance/{0}", instance);
```
we prefer always use square brackets  on if conditions

unaizorrilla commented on src/HealthChecks.Publisher.Prometheus/PrometheusGatewayPublisher.cs:
```diff
@@ -0,0 +1,66 @@
+            if (!string.IsNullOrEmpty(instance)) sb.AppendFormat("/instance/{0}", instance);
+
+            if (!Uri.TryCreate(sb.ToString(), UriKind.Absolute, out _targetUrl))
+                throw new ArgumentException("Endpoint must be a valid url", nameof(endpoint));
```
we use always square brackets on if conditions

unaizorrilla commented on src/HealthChecks.Publisher.Prometheus/PrometheusResponseWriter.cs:
```diff
@@ -0,0 +1,27 @@
+        {
+            var instance = new PrometheusResponseWriter();
+            instance.WriteMetricsFromHealthReport(report);
+            var resultStream = CollectionToStreamWriter(instance.Registry);
```
Dispose used streams

unaizorrilla commented:
Hi @SbiCA 

Looks great in general, please fix minor issues

SbiCA commented:
Hi @unaizorrilla 

No worries, I fixed the things you commented and also adjusted the build.ps1 so we should be able to get a package published let me know if there is anything else left.

Cheers

unaizorrilla commented:
I try to review this in a few minutes :-)
